CONTRIBUTING.md

```diff
@@ -1,73 +1,78 @@
+4. Thou shalt properly test thy code. This includes unit *and* integration tests. Try to isolate parts that can be individually tested.
+
+# C++
+* Use automatic memory management (RAII, smart pointers). `new` and `malloc` are evil words.
+* Be mindful of ownership. Not everything needs to be a smart pointer. Consider passing around references to the object or references to a shared_ptr instead of copying the shared_ptr. Remember that this might not be safe when passing shared_ptrs into, i.e., JobTasks.
+* Use `const` whenever possible. Consider variables, methods, pointers, and their pointees.
+* Reduce the size of your hpp files, both in terms of the number of lines and the code complexity. This keeps the compilation duration low.
+  * Code in hpp files is compiled for every cpp file that includes the header. As such, move code to cpp files where possible. This often includes templated classes, where it is sometimes possible to implement their code in cpp files.
+  * Anonymous namespaces are a good way to define local helper methods.
+  * Use forward declarations instead of full header includes wherever possible.
+* Loops
+  * Use range-based for loops when possible: `for (const auto& item : items) {...}`.
+  * If you have to use old-style loops, keep in mind that the loop condition is evaluated every time: Instead of `for (auto offset = size_t{0}; offset < something.size(); ++offset)`, the size should be retrieved just once. See also [here](http://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop)
+* Data structures
+  * When creating a vector where you know the size beforehand, use `reserve` to avoid unnecessary resizes and allocations.
+  * Hash-based data structures are usually faster than tree-based data structures. Unless you have a reason to use the latter, prefer unordered_(map|set) over map and set.
+* Copies
+
+    * C++ makes it too easy to inadvertently copy objects.
+
+    * For larger elements (e.g., vectors), pass a (const) reference instead.
+
+    * When your implemented class does not need to be copied (e.g., a Table should never exist more than once), inherit from Noncopyable to avoid these potentially expensive copies.
+
+* Miscellaneous
+  * Prefer `if (object) {` over `if (object != nullptr) {` or `if (object.has_value()) {`.
+  * Don't write `this->` if you don't have to.
+  * Be explicit with types: Use [u]int(8|16|32|64)_t instead of `int, long, uint` etc.
+  * Use [auto-to-stick](https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/): `auto x = 17;` or `auto y = std::vector<size_t>{};`.
+  * Namespaces: Do not create nested namespaces, do not import namespaces.
+  * Consider structured bindings: `const auto& [iterator, added] = unordered_map.emplace(...);`
+
```

Comment:
Add some notes on templated methods? Maybe here or add another bullet point. 

These are the notes from the previous guidelines:

> If your templated methods/classes are used outside of your file, they have to be in the header. But if you only use them internally, you should place them in the cpp file.

> Try to keep the size of your templated methods as small as possible (Thin Template idiom). If only a small part of your method depends on the template parameter, consider moving the rest into a non-templated method. This reduces compile times.

Reply:
Move Thin Template to Ressource Page